src/chain/chain_state.cpp

```diff
@@ -317,8 +321,14 @@ uint32_t chain_state::work_required_retarget(const data& values,
     BITCOIN_ASSERT_MSG(!bits.is_overflowed(), "previous block has bad bits");
 
     uint256_t target(bits);
+    const bool needs_shift = msb(target) > msb(pow_limit) - 1;
```

Comment:
I should use `auto` here.

Reply:
Incorporate flag into this result:

```cpp
const auto patch = script::is_enabled(forks, rule_fork::retarget_overflow_patch);
const auto shift = patch && (msb(target) > msb(pow_limit) - 1);
```

This is simpler code, prevents one condition and four math operations if the flag is not enabled.

is this safe from underflow?
```
msb(pow_limit) - 1
```

What namespace and header is `msb`?

Reply:
> is this safe from underflow?

No, it's not. But that's a big edge case since it'll only underflow when the user chooses `bc::settings.proof_of_work_limit` to represent `uint256_t(1)` (e.g. 0x1010000). I've never seen a block that met this difficulty requirement :wink:. However, it's trivial to fix the underflow so I'll do it.

Reply:
Isn’t it a consensus question?

Reply:
This is not easy to answer: if we take ltc or ftc then this will never underflow for what they've chosen as `bc::settings.proof_of_work_limit`. Moreover, what you can see here is exactly what they've implemented. But nowadays we have so many coins - I'm not able to answer this for all of them :man_shrugging:

Reply:
Okay, makes sense. Pretty obvious, just didn't look too closely. I would use this then:
```cpp
const auto shift = patch && (msb(target) > safe_subtract(msb(pow_limit), 1) ? 1u : 0u;
```
This will throw if it underflows. FWIW we try to never allow potential under/overflow without explicit handling (or a comment as to why it is correct behavior). So any time I see math I question it.